{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "%matplotlib inline\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "DPI = 300\n",
    "FONTSIZE = 20  # 42\n",
    "\n",
    "sc.settings.set_figure_params(\n",
    "    scanpy=True, dpi=100, transparent=True, vector_friendly=True, dpi_save=DPI\n",
    ")\n",
    "from matplotlib import rcParams\n",
    "\n",
    "rcParams[\"pdf.fonttype\"] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIR2SAVE = Path(\n",
    "    \"/data/BCI-CRC/nasrine/data/CRC/Primary_CRC_dataset/subpopulations/TNKILC\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FIG2SAVE = DIR2SAVE.joinpath(\"figures/\")\n",
    "FIG2SAVE\n",
    "# set the global variable: sc.settings.figdir to save all plots\n",
    "sc.settings.figdir = FIG2SAVE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\n",
    "    DIR2SAVE.joinpath(\"SMC_KUL_Pelka_Che_Wu_CRC_integrated_scvi_hvg_TNKILC.h5ad\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Leiden clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[\"neighbors\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# neighbors were already computed using Harmony corrected PCs, so we perform clustering on that graph\n",
    "sc.tl.leiden(adata, key_added=\"leiden_scVI\", resolution=1, random_state=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"leiden_scVI\",\n",
    "    legend_loc=\"on data\",\n",
    "    save=\"general_clustering.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load cell cycle scores saved from integration\n",
    "cell_cycle_scores = pd.read_csv(\n",
    "    \"/data/BCI-CRC/nasrine/data/CRC/Primary_CRC_dataset/data_integration/SMC_KUL_Pelka_Che_Wu_CRC_integrated_scvi_hvg_cc_scores.txt\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = adata.obs.merge(\n",
    "    cell_cycle_scores, how=\"left\", left_index=True, right_index=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\"S_score\", \"G2M_score\", \"phase\", \"cell_cycle_diff\"],\n",
    "    color_map=\"viridis\",\n",
    "    save=\"cell_cycle.pdf\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\"doublet_score\", \"n_genes_by_counts\", \"pct_counts_mt\", \"pct_counts_ribo\"],\n",
    "    color_map=\"viridis\",\n",
    "    save=\"QC_covariates.pdf\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at distribution of cell source across clusters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.patches import Rectangle\n",
    "\n",
    "\n",
    "def proportion_cells_patient(\n",
    "    adata, groupby_labels, xlabel: str, ylabel: str, colors: dict\n",
    "):  # colors\n",
    "    # compute proportion of cells within each group\n",
    "    table2plot = (\n",
    "        adata.reset_index()\n",
    "        .groupby(groupby_labels)\n",
    "        .size()\n",
    "        .groupby(level=0)\n",
    "        .apply(lambda x: x * 100 / x.sum())\n",
    "        .unstack()\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(\n",
    "        nrows=1, ncols=1, sharey=False, sharex=False, dpi=DPI, figsize=(4, 4)\n",
    "    )\n",
    "\n",
    "    print(table2plot)\n",
    "    table2plot.plot.bar(stacked=True, ax=ax, color=colors.values())  # , color=colors\n",
    "    ax.set_ylabel(ylabel)\n",
    "    ax.set_xlabel(xlabel)\n",
    "    ax.grid(False, which=\"major\", axis=\"both\")  # removes major horizontal gridlinesd\n",
    "\n",
    "    labels = list(colors.keys())\n",
    "    l = [Rectangle((0, 0), 0, 0, color=color) for color in list(colors.values())]\n",
    "    ax.legend(\n",
    "        l,\n",
    "        labels,\n",
    "        loc=\"upper left\",\n",
    "        bbox_to_anchor=(1, 0.8),\n",
    "        facecolor=\"white\",\n",
    "        edgecolor=\"white\",\n",
    "        ncol=1,\n",
    "        borderaxespad=0.0,\n",
    "        framealpha=0,\n",
    "        frameon=False,\n",
    "    )\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import OrderedDict\n",
    "\n",
    "color_dict = OrderedDict(\n",
    "    zip(\n",
    "        adata.obs[\"cell_source\"].cat.categories.values.tolist(),\n",
    "        adata.uns[\"cell_source_colors\"],\n",
    "    )\n",
    ")\n",
    "\n",
    "proportion_cells_patient(\n",
    "    adata.obs,\n",
    "    groupby_labels=[\"leiden_scVI\", \"cell_source\"],\n",
    "    xlabel=\"leiden_scVI\",\n",
    "    ylabel=\"Percent cell source\",\n",
    "    colors=color_dict,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot some markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_T_general = {\n",
    "    \"T\": [\"TRAC\"],\n",
    "    \"CD4\": [\"CD4\"],\n",
    "    \"CD8\": [\"CD8A\", \"CD8B\"],\n",
    "    \"Cycling\": [\"MKI67\"],\n",
    "    \"Treg\": [\"FOXP3\", \"CTLA4\", \"CCR4\", \"IL2RA\"],  # Treg are CD127 (IL7R) low.\n",
    "    \"Exhausted T\": [\n",
    "        \"PDCD1\",\n",
    "        \"LAG3\",\n",
    "        \"HAVCR2\",\n",
    "        \"CTLA4\",\n",
    "        \"TIGIT\",\n",
    "        \"ENTPD1\",\n",
    "    ],  # PD1: PCDC1, TIM3: HAVCR2,  CD39:ENTPD1 https://www.nature.com/articles/s41467-021-23324-4\n",
    "    \"NK\": [\n",
    "        \"EOMES\",\n",
    "        \"CMC1\",\n",
    "        \"GZMK\",\n",
    "        \"XCL1\",\n",
    "        \"NKG7\",\n",
    "        \"PRF1\",\n",
    "        \"NCR1\",\n",
    "        \"NCAM1\",\n",
    "        \"FCER1G\",\n",
    "        \"ITGA1\",\n",
    "        \"GZMB\",\n",
    "        \"FCGR3A\",\n",
    "    ],\n",
    "    \"ILC\": [\"AREG\", \"TLE1\", \"IL4I1\"],\n",
    "}\n",
    "\n",
    "markers_T_cd4 = {\n",
    "    \"Th\": [\"CD4\", \"TRAC\", \"CD3D\", \"TRBC1\", \"TRBC2\"],\n",
    "    \"Tfh\": [\"ICOS\", \"CXCR5\", \"TCF7\", \"PDCD1\", \"CCR7\"],\n",
    "    \"Naïve CD4 T\": [\n",
    "        \"CCR7\",\n",
    "        \"SELL\",\n",
    "        \"TCF7\",\n",
    "        \"LEF1\",\n",
    "    ],  # https://www.nature.com/articles/s41467-019-12464-3\n",
    "    \"Th1/Th17/Th2\": [\n",
    "        \"CXCR3\",\n",
    "        \"TBX21\",\n",
    "        \"CCL5\",\n",
    "        \"CCR6\",\n",
    "        \"IL22\",\n",
    "        \"RORA\",\n",
    "        \"IL7R\",\n",
    "        \"IL4\",\n",
    "        \"IL13\",\n",
    "        \"GATA3\",\n",
    "        \"CCR4\",\n",
    "    ],\n",
    "    \"Th17\": [\n",
    "        \"IL17A\",\n",
    "        \"ODF2L\",\n",
    "        \"IL7R\",\n",
    "        \"PDE4D\",\n",
    "    ],  #'CCR4', 'CCR6', 'IL1R1', 'IL6R', 'IL21R', 'IL23R'],# 'TGFBR1', 'RORA', 'RORC', 'BATF', 'IRF4'], # IL17 not in data\n",
    "    \"Th1\": [\"CCL5\", \"PHLDA1\", \"LYAR\"],\n",
    "}\n",
    "\n",
    "markers_T_cd8 = {\n",
    "    \"Effector CD8\": [\n",
    "        \"CCL4\",\n",
    "        \"CCL5\",\n",
    "        \"GZMK\",\n",
    "        \"GZMB\",\n",
    "        \"PFN1\",\n",
    "        \"GZMA\",\n",
    "        \"GZMH\",\n",
    "        \"NKG7\",\n",
    "    ],  # Cytotoxic is same as effector for the effector, if it’s only one cluster, and doesn’t have TCF7/CCR7, I would annotate them just as effector\n",
    "    \"Tmem\": [\n",
    "        \"CCR7\",\n",
    "        \"PTPRC\",\n",
    "        \"ENPP1\",\n",
    "    ],  # https://panglaodb.se/markers.html?cell_type=%27T%20memory%20cells%27\n",
    "    \"Naive cytotoxic\": [\"CD8A\", \"CCR7\", \"SELL\"],\n",
    "}\n",
    "\n",
    "markers_T_other = {\n",
    "    \"gdT\": [\"KLRC2\", \"TRGC1\", \"TRGC2\", \"TRDC\"],  # not in data: 'TCRD','TCRG'\n",
    "    \"NKT\": [\"GZMA\", \"CCL5\", \"NKG7\", \"KLRB1\", \"CD3G\", \"FGFBP2\"],\n",
    "    \"MAIT\": [\"SLC4A10\", \"NCR3\", \"KLRB1\"],\n",
    "}\n",
    "\n",
    "markers_ILC = {\n",
    "    \"ILC\": [\n",
    "        \"AREG\",\n",
    "        \"TLE1\",\n",
    "        \"IL4I1\",\n",
    "    ],  # ['NCR2', 'ITGAE', 'KIT', 'IL7R', 'KLRB1', 'AHR'],\n",
    "    \"ILC1\": [\"TBX21\", \"CD3D\", \"CXCR3\", \"PLCD4\"],\n",
    "    \"ILC2\": [\"KRT1\", \"HPGDS\", \"SLAMF1\"],  # ['HPGDS', 'GATA3', 'PTGDR2', 'IL1RL1'],\n",
    "    \"ILC3\": [\"IL4I1\", \"RORC\", \"TNFRSF25\", \"SPINK2\", \"KLRB1\", \"IL7R\"],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "# use log1p data stored in .raw\n",
    "markers2plot = list(\n",
    "    itertools.chain(*list(markers_T_general.values()))\n",
    ")  # get all markers in a single list\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=markers2plot,\n",
    "    use_raw=True,\n",
    "    vmin=0.0,\n",
    "    vmax=\"p99\",\n",
    "    color_map=\"plasma_r\",  #'RdPu',\n",
    "    save=\"general_markers.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "# use log1p data stored in .raw\n",
    "markers2plot = list(\n",
    "    itertools.chain(*list(markers_T_cd4.values()))\n",
    ")  # get all markers in a single list\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=markers2plot,\n",
    "    use_raw=True,\n",
    "    vmin=0.0,\n",
    "    vmax=\"p99\",\n",
    "    color_map=\"plasma_r\",  #'RdPu',\n",
    "    save=\"general_markers_cd4.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "# use log1p data stored in .raw\n",
    "markers2plot = list(\n",
    "    itertools.chain(*list(markers_T_cd8.values()))\n",
    ")  # get all markers in a single list\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=markers2plot,\n",
    "    use_raw=True,\n",
    "    vmin=0.0,\n",
    "    vmax=\"p99\",\n",
    "    color_map=\"plasma_r\",  #'RdPu',\n",
    "    save=\"general_markers_cd8.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "# use log1p data stored in .raw\n",
    "markers2plot = list(\n",
    "    itertools.chain(*list(markers_T_other.values()))\n",
    ")  # get all markers in a single list\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=markers2plot,\n",
    "    use_raw=True,\n",
    "    vmin=0.0,\n",
    "    vmax=\"p99\",\n",
    "    color_map=\"plasma_r\",  #'RdPu',\n",
    "    save=\"general_markers_other.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "# use log1p data stored in .raw\n",
    "markers2plot = list(\n",
    "    itertools.chain(*list(markers_ILC.values()))\n",
    ")  # get all markers in a single list\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=markers2plot,\n",
    "    use_raw=True,\n",
    "    vmin=0.0,\n",
    "    vmax=\"p99\",\n",
    "    color_map=\"plasma_r\",  #'RdPu',\n",
    "    save=\"general_markers_ILC.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### markers for stress\n",
    "# HSP\n",
    "dissocation_markers_dict = {\n",
    "    \"shock protein\": [\n",
    "        \"HSPE1\",\n",
    "        \"HSPA1A\",\n",
    "        \"HSPA1B\",\n",
    "        \"HSP90AA1\",\n",
    "        \"HSP90AB1\",\n",
    "        \"HSPA8\",\n",
    "        \"HSPB1\",\n",
    "    ],\n",
    "    \"immediate early genes\": [\"FOS\", \"JUN\"],\n",
    "}\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=dissocation_markers_dict[\"shock protein\"],\n",
    "    vmax=\"p99\",\n",
    "    use_raw=True,\n",
    "    vmin=0,\n",
    "    color_map=\"plasma_r\",\n",
    "    save=\"general_HSPmarkers.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential expression to get DE genes upregulated per cluster "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# issue here https://github.com/theislab/single-cell-tutorial/issues/97\n",
    "# This seems to be a scanpy bug as you can see here and here. The latter issue suggests to just add the line:\n",
    "# adata.uns['log1p'][\"base\"] = None after reading again, or downgrading to AnnData<0.8.\n",
    "# Either way, this should be fixed soon by the maintenance team.\n",
    "adata.uns[\"log1p\"][\"base\"] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    reference=\"rest\",\n",
    "    method=\"wilcoxon\",\n",
    "    use_raw=True,\n",
    "    layer=None,\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    key=\"rank_genes_wilcoxon\",\n",
    "    var_names=markers_T_general,\n",
    "    values_to_plot=\"logfoldchanges\",\n",
    "    cmap=\"bwr\",\n",
    "    vmin=-4,\n",
    "    vmax=4,\n",
    "    min_logfoldchange=1,\n",
    "    colorbar_title=\"log fold change\",\n",
    "    save=\"general_dotplot.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    key=\"rank_genes_wilcoxon\",\n",
    "    var_names=markers_T_cd8,\n",
    "    values_to_plot=\"logfoldchanges\",\n",
    "    cmap=\"bwr\",\n",
    "    vmin=-4,\n",
    "    vmax=4,\n",
    "    min_logfoldchange=1,\n",
    "    colorbar_title=\"log fold change\",\n",
    "    save=\"general_dotplot_cd8.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    key=\"rank_genes_wilcoxon\",\n",
    "    var_names=markers_T_other,\n",
    "    values_to_plot=\"logfoldchanges\",\n",
    "    cmap=\"bwr\",\n",
    "    vmin=-4,\n",
    "    vmax=4,\n",
    "    min_logfoldchange=1,\n",
    "    colorbar_title=\"log fold change\",\n",
    "    save=\"general_dotplot_other.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    use_raw=True,\n",
    "    var_names=markers_T_other,\n",
    "    cmap=\"plasma_r\",\n",
    "    standard_scale=\"var\",\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    "    colorbar_title=\"Mean expression\",\n",
    "    dendrogram=False,\n",
    "    save=\"TNKILC_dotplot_mean_Tother.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    key=\"rank_genes_wilcoxon\",\n",
    "    var_names=dissocation_markers_dict,\n",
    "    values_to_plot=\"logfoldchanges\",\n",
    "    cmap=\"bwr\",\n",
    "    vmin=-4,\n",
    "    vmax=4,\n",
    "    min_logfoldchange=1,\n",
    "    colorbar_title=\"log fold change\",\n",
    "    save=\"general_dotplot_HSP.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    key=\"rank_genes_wilcoxon\",\n",
    "    var_names=markers_T_cd4,\n",
    "    values_to_plot=\"logfoldchanges\",\n",
    "    cmap=\"bwr\",\n",
    "    vmin=-4,\n",
    "    vmax=4,\n",
    "    min_logfoldchange=1,\n",
    "    colorbar_title=\"log fold change\",\n",
    "    save=\"general_dotplot_cd4.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    use_raw=True,\n",
    "    var_names=markers_T_general,\n",
    "    cmap=\"plasma_r\",\n",
    "    standard_scale=\"var\",\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    "    colorbar_title=\"Mean expression\",\n",
    "    dendrogram=False,\n",
    "    save=\"TNKILC_dotplot_mean.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "sc.pl.dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    use_raw=True,\n",
    "    var_names=markers_T_cd4,\n",
    "    cmap=\"plasma_r\",\n",
    "    standard_scale=\"var\",\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    "    colorbar_title=\"Mean expression\",\n",
    "    dendrogram=False,\n",
    "    save=\"TNKILC_CD4_dotplot_mean.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save intermediate result\n",
    "# write to file\n",
    "adata.write(\n",
    "    DIR2SAVE.joinpath(\n",
    "        \"SMC_KUL_Pelka_Che_Wu_CRC_integrated_scvi_hvg_TNKILC_clustering.h5ad\"\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pval_thresh = 0.05\n",
    "log2fc_thresh = 1\n",
    "cluster_de_genes = dict()\n",
    "for cluster in sorted(set(adata.obs[\"leiden_scVI\"])):\n",
    "    cluster_de_genes[cluster] = sc.get.rank_genes_groups_df(\n",
    "        adata,\n",
    "        group=cluster,\n",
    "        key=\"rank_genes_wilcoxon\",\n",
    "        pval_cutoff=pval_thresh,\n",
    "        log2fc_min=log2fc_thresh,\n",
    "        log2fc_max=None,\n",
    "    ).sort_values(\"logfoldchanges\", ascending=False)\n",
    "\n",
    "# write to excel file DE genes per cluster\n",
    "# Create a Pandas Excel writer using XlsxWriter as the engine.\n",
    "path2save = DIR2SAVE.joinpath(\n",
    "    \"TNKILC_pval{}_log2fc{}.xlsx\".format(pval_thresh, log2fc_thresh)\n",
    ")\n",
    "\n",
    "with pd.ExcelWriter(path2save) as writer:\n",
    "    for cluster in list(cluster_de_genes.keys()):\n",
    "\n",
    "        # get celltype of cluster\n",
    "        # celltype = np.unique(adata[adata.obs['leiden']==cluster,:].obs['cell identity'])[0]\n",
    "\n",
    "        cluster_de_genes[cluster].to_excel(\n",
    "            writer, sheet_name=\"cluster{}\".format(cluster)\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at scores for general signatures "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# t phenotype\n",
    "### see some signatures from Zhang paper\n",
    "# read excel file\n",
    "tcell_phen_zhang = pd.read_excel(\n",
    "    \"/data/BCI-CRC/nasrine/data/zhang_signatures.xlsx\",\n",
    "    sheet_name=\"Tcell_markers\",\n",
    "    skiprows=None,\n",
    "    header=1,\n",
    "    dtype=str,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(set(tcell_phen_zhang.Cluster))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NKT_signature = tcell_phen_zhang[\n",
    "    tcell_phen_zhang[\"Cluster\"].str.contains(\"c30_NKT-FCGR3A\")\n",
    "][\"Gene\"].values.tolist()\n",
    "print(len(NKT_signature))\n",
    "NKT_signature_var = [\n",
    "    gene for gene in NKT_signature if gene in adata.raw.to_adata().var_names\n",
    "]\n",
    "sc.tl.score_genes(\n",
    "    adata, gene_list=NKT_signature_var, score_name=\"NKT_score\", use_raw=True\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"NKT_score\",\n",
    "    vmax=\"p99\",\n",
    "    use_raw=True,\n",
    "    vmin=0,\n",
    "    save=\"TNKILC_NKT_score.pdf\",\n",
    "    color_map=\"YlOrRd\",\n",
    "    show=True,\n",
    ")\n",
    "print(\"length of gene list in adata.var:\", len(NKT_signature_var))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CD8_HSP_signature = tcell_phen_zhang[\n",
    "    tcell_phen_zhang[\"Cluster\"].str.contains(\"c12_CD8-HSPA1A\")\n",
    "][\"Gene\"].values.tolist()\n",
    "print(len(CD8_HSP_signature))\n",
    "CD8_HSP_signature_var = [\n",
    "    gene for gene in CD8_HSP_signature if gene in adata.raw.to_adata().var_names\n",
    "]\n",
    "sc.tl.score_genes(\n",
    "    adata, gene_list=CD8_HSP_signature_var, score_name=\"CD8_HSP_score\", use_raw=True\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"CD8_HSP_score\",\n",
    "    vmax=\"p99\",\n",
    "    use_raw=True,\n",
    "    vmin=0,\n",
    "    save=\"TNKILC_CD8_HSP_score.pdf\",\n",
    "    color_map=\"YlOrRd\",\n",
    "    show=True,\n",
    ")\n",
    "print(\"length of gene list in adata.var:\", len(CD8_HSP_signature_var))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Trm_XCL1_signature = tcell_phen_zhang[\n",
    "    tcell_phen_zhang[\"Cluster\"].str.contains(\"C09_CD8_Trm-XCL1\")\n",
    "][\"Gene\"].values.tolist()\n",
    "print(len(Trm_XCL1_signature))\n",
    "Trm_XCL1_signature_var = [\n",
    "    gene for gene in Trm_XCL1_signature if gene in adata.raw.to_adata().var_names\n",
    "]\n",
    "sc.tl.score_genes(\n",
    "    adata, gene_list=Trm_XCL1_signature_var, score_name=\"Trm_XCL1_score\", use_raw=True\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"Trm_XCL1_score\",\n",
    "    vmax=\"p99\",\n",
    "    use_raw=True,\n",
    "    vmin=0,\n",
    "    save=\"TNKILC_Trm_XCL1_score.pdf\",\n",
    "    color_map=\"YlOrRd\",\n",
    "    show=True,\n",
    ")\n",
    "print(\"length of gene list in adata.var:\", len(Trm_XCL1_signature_var))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CD8_Te_GZMK_signature = tcell_phen_zhang[\n",
    "    tcell_phen_zhang[\"Cluster\"].str.contains(\"C05_CD8_Tem-GZMK\")\n",
    "][\"Gene\"].values.tolist()\n",
    "print(len(CD8_Te_GZMK_signature))\n",
    "CD8_Te_GZMK_signature_var = [\n",
    "    gene for gene in CD8_Te_GZMK_signature if gene in adata.raw.to_adata().var_names\n",
    "]\n",
    "sc.tl.score_genes(\n",
    "    adata,\n",
    "    gene_list=CD8_Te_GZMK_signature_var,\n",
    "    score_name=\"CD8_Te_GZMK_score\",\n",
    "    use_raw=True,\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"CD8_Te_GZMK_score\",\n",
    "    vmax=\"p99\",\n",
    "    use_raw=True,\n",
    "    vmin=0,\n",
    "    save=\"TNKILC_CD8_Te_GZMK_score.pdf\",\n",
    "    color_map=\"YlOrRd\",\n",
    "    show=True,\n",
    ")\n",
    "print(\"length of gene list in adata.var:\", len(CD8_Te_GZMK_signature_var))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CD4_Tn_signature = tcell_phen_zhang[tcell_phen_zhang[\"Cluster\"].str.contains(\"CD4_Tn\")][\n",
    "    \"Gene\"\n",
    "].values.tolist()\n",
    "print(len(CD4_Tn_signature))\n",
    "CD4_Tn_signature_var = [\n",
    "    gene for gene in CD4_Tn_signature if gene in adata.raw.to_adata().var_names\n",
    "]\n",
    "sc.tl.score_genes(\n",
    "    adata, gene_list=CD4_Tn_signature_var, score_name=\"CD4_Tn_score\", use_raw=True\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"CD4_Tn_score\",\n",
    "    vmax=\"p99\",\n",
    "    use_raw=True,\n",
    "    vmin=0,\n",
    "    save=\"TNKILC_CD4_Tn_score.pdf\",\n",
    "    color_map=\"YlOrRd\",\n",
    "    show=True,\n",
    ")\n",
    "print(\"length of gene list in adata.var:\", len(CD4_Tn_signature_var))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CD4_Treg_signature = tcell_phen_zhang[\n",
    "    tcell_phen_zhang[\"Cluster\"].str.contains(\"CD4_Treg\")\n",
    "][\"Gene\"].values.tolist()\n",
    "print(len(CD4_Treg_signature))\n",
    "CD4_Treg_signature_var = [\n",
    "    gene for gene in CD4_Treg_signature if gene in adata.raw.to_adata().var_names\n",
    "]\n",
    "sc.tl.score_genes(\n",
    "    adata, gene_list=CD4_Treg_signature_var, score_name=\"CD4_Treg_score\", use_raw=True\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"CD4_Treg_score\",\n",
    "    vmax=\"p99\",\n",
    "    use_raw=True,\n",
    "    vmin=0,\n",
    "    save=\"TNKILC_CD4_Treg_score.pdf\",\n",
    "    color_map=\"YlOrRd\",\n",
    "    show=True,\n",
    ")\n",
    "print(\"length of gene list in adata.var:\", len(CD4_Treg_signature_var))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compare DE genes\n",
    "# DE genes between Therapy treated vs None for C1QC TAMs\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    groups=[\"3\"],\n",
    "    reference=\"5\",\n",
    "    method=\"wilcoxon\",\n",
    "    layer=\"log1p\",\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon_3\",\n",
    "    use_raw=False,\n",
    ")\n",
    "\n",
    "sc.pl.rank_genes_groups(adata, groups=[\"3\"], n_genes=30, key=\"rank_genes_wilcoxon_3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compare DE genes\n",
    "# DE genes between Therapy treated vs None for C1QC TAMs\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    groups=[\"5\"],\n",
    "    reference=\"3\",\n",
    "    method=\"wilcoxon\",\n",
    "    layer=\"log1p\",\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon_5\",\n",
    "    use_raw=False,\n",
    ")\n",
    "\n",
    "sc.pl.rank_genes_groups(adata, groups=[\"5\"], n_genes=30, key=\"rank_genes_wilcoxon_5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compare DE genes\n",
    "# DE genes between Therapy treated vs None for C1QC TAMs\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    groups=[\"2\"],\n",
    "    reference=\"10\",\n",
    "    method=\"wilcoxon\",\n",
    "    layer=\"log1p\",\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon_2\",\n",
    "    use_raw=False,\n",
    ")\n",
    "\n",
    "sc.pl.rank_genes_groups(adata, groups=[\"2\"], n_genes=30, key=\"rank_genes_wilcoxon_2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compare DE genes\n",
    "# DE genes between Therapy treated vs None for C1QC TAMs\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    groups=[\"10\"],\n",
    "    reference=\"2\",\n",
    "    method=\"wilcoxon\",\n",
    "    layer=\"log1p\",\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon_10\",\n",
    "    use_raw=False,\n",
    ")\n",
    "\n",
    "sc.pl.rank_genes_groups(adata, groups=[\"10\"], n_genes=30, key=\"rank_genes_wilcoxon_10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compare DE genes\n",
    "# DE genes between Therapy treated vs None for C1QC TAMs\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    groups=[\"10\"],\n",
    "    reference=\"13\",\n",
    "    method=\"wilcoxon\",\n",
    "    layer=\"log1p\",\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon_10_13\",\n",
    "    use_raw=False,\n",
    ")\n",
    "\n",
    "sc.pl.rank_genes_groups(\n",
    "    adata, groups=[\"10\"], n_genes=30, key=\"rank_genes_wilcoxon_10_13\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compare DE genes\n",
    "# DE genes between Therapy treated vs None for C1QC TAMs\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    groups=[\"13\"],\n",
    "    reference=\"10\",\n",
    "    method=\"wilcoxon\",\n",
    "    layer=\"log1p\",\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon_13_10\",\n",
    "    use_raw=False,\n",
    ")\n",
    "\n",
    "sc.pl.rank_genes_groups(\n",
    "    adata, groups=[\"13\"], n_genes=30, key=\"rank_genes_wilcoxon_13_10\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compare DE genes\n",
    "# DE genes between Therapy treated vs None for C1QC TAMs\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    groups=[\"13\"],\n",
    "    reference=\"2\",\n",
    "    method=\"wilcoxon\",\n",
    "    layer=\"log1p\",\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon_13_2\",\n",
    "    use_raw=False,\n",
    ")\n",
    "\n",
    "sc.pl.rank_genes_groups(\n",
    "    adata, groups=[\"13\"], n_genes=30, key=\"rank_genes_wilcoxon_13_2\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compare DE genes\n",
    "# DE genes between Therapy treated vs None for C1QC TAMs\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    groups=[\"2\"],\n",
    "    reference=\"13\",\n",
    "    method=\"wilcoxon\",\n",
    "    layer=\"log1p\",\n",
    "    pts=True,\n",
    "    corr_method=\"benjamini-hochberg\",\n",
    "    key_added=\"rank_genes_wilcoxon_2_13\",\n",
    "    use_raw=False,\n",
    ")\n",
    "\n",
    "sc.pl.rank_genes_groups(adata, groups=[\"2\"], n_genes=30, key=\"rank_genes_wilcoxon_2_13\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\n",
    "    DIR2SAVE.joinpath(\n",
    "        \"SMC_KUL_Pelka_Che_Wu_CRC_integrated_scvi_hvg_TNKILC_clustering.h5ad\"\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_to_new = {\n",
    "    \"0\": \"CD8 Tex\",  # exhausted GZMB, GZMA, CCL5, HAVCR2, GNLY, NKG7, CCL4, CD8A, IFNG, GZMH, LAG3, PRF1, TIGIT, PDCD1, TOX resembles mirjana T cell paper TPEX in Two subsets of stem-like CD8 memory T cell progenitors with 2 distinct fate commitments in humans\n",
    "    \"1\": \"CD4 Tn\",  # naive CCR7, LEF1, SELL, TCF7\n",
    "    \"2\": \"CD8 Tem\",  # effector memory GZMK, CCL4, CD8A, GZMM, HLA-DRB1, HLA-DRA, IFNG, CCL5, NKG7, EOMES, GZMA\n",
    "    \"3\": \"Treg\",  # FOXP3, IL2RA, CTLA4, TIGIT\n",
    "    \"4\": \"CD4 Th\",  # ANXA1 , CD40LG, IL7R\n",
    "    \"5\": \"Treg HSP\",  # FOXP3, IL2RA, CTLA4, TIGIT, CD79B?\n",
    "    \"6\": \"NK\",  # NCAM1, FCER1G, KLRC1\n",
    "    \"7\": \"CD4 Th17\",  # Il22, IL17A, RORA, PDE4D\n",
    "    \"8\": \"T cycling\",\n",
    "    \"9\": \"gdT\",  # TRDC, TRGC1, ITGA1 but has a bit of NK maybe\n",
    "    \"10\": \"Myeloid-T doublet\",  # APOC1, SPP1, APOE, HSP, IGHG, JCHAIN\n",
    "    \"11\": \"CD4 Th HSP\",  # CD40LG, IL7R plus umap has cd4\n",
    "    \"12\": \"CD4 Tfh\",  # CD4, ITM2A, LPAR6, PDCD1\n",
    "    \"13\": \"CD8 Tem HSP\",  # effector memory HSP, GZMK, CCL5, CCL4, IFNG, EOMES, NKG7, GZMH, CD8B, CD8A, GZMA\n",
    "    \"14\": \"B-T doublet\",  # CD22, CD79A, MS4A1, CD79B\n",
    "    \"15\": \"ILC3\",  # IL4I1, RORC, TNFRSF25\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"Annotation_scVI_detailed\"] = (\n",
    "    adata.obs[\"leiden_scVI\"].map(old_to_new).astype(\"category\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata, color=\"Annotation_scVI_detailed\", show=True, legend_loc=\"right margin\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# todo dotplot with all cell types annotated\n",
    "markers_T_dotplot = {\n",
    "    \"T\": [\"TRAC\"],\n",
    "    \"CD4\": [\"CD4\"],\n",
    "    \"CD8\": [\"CD8A\", \"CD8B\"],\n",
    "    \"Cycling\": [\"MKI67\"],\n",
    "    \"Treg\": [\"FOXP3\", \"CTLA4\", \"CCR4\", \"IL2RA\"],  # Treg are CD127 (IL7R) low.\n",
    "    \"Exhausted T\": [\n",
    "        \"PDCD1\",\n",
    "        \"LAG3\",\n",
    "        \"HAVCR2\",\n",
    "        \"CTLA4\",\n",
    "        \"TIGIT\",\n",
    "        \"ENTPD1\",\n",
    "    ],  # PD1: PCDC1, TIM3: HAVCR2,  CD39:ENTPD1 https://www.nature.com/articles/s41467-021-23324-4\n",
    "    \"NK\": [\n",
    "        \"EOMES\",\n",
    "        \"CMC1\",\n",
    "        \"GZMK\",\n",
    "        \"XCL1\",\n",
    "        \"NKG7\",\n",
    "        \"PRF1\",\n",
    "        \"NCR1\",\n",
    "        \"NCAM1\",\n",
    "        \"FCER1G\",\n",
    "        \"ITGA1\",\n",
    "        \"GZMB\",\n",
    "        \"FCGR3A\",\n",
    "    ],\n",
    "    \"ILC\": [\"AREG\", \"TLE1\", \"IL4I1\"],\n",
    "    \"Th\": [\"CD4\", \"TRAC\", \"CD3D\", \"TRBC1\", \"TRBC2\"],\n",
    "    \"Tfh\": [\"ICOS\", \"CXCR5\", \"TCF7\", \"PDCD1\", \"CCR7\"],\n",
    "    \"Naïve CD4 T\": [\"CCR7\", \"SELL\", \"TCF7\", \"LEF1\"],\n",
    "    \"Th17\": [\"IL17A\", \"ODF2L\", \"IL7R\", \"PDE4D\"],\n",
    "    \"Effector CD8\": [\"CCL4\", \"CCL5\", \"GZMK\", \"GZMB\", \"PFN1\", \"GZMA\", \"GZMH\", \"NKG7\"],\n",
    "    \"gdT\": [\"KLRC2\", \"TRGC1\", \"TRGC2\", \"TRDC\"],\n",
    "    \"ILC3\": [\"IL4I1\", \"RORC\", \"TNFRSF25\", \"SPINK2\", \"KLRB1\", \"IL7R\"],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    use_raw=True,\n",
    "    var_names=markers_T_dotplot,\n",
    "    cmap=\"plasma_r\",\n",
    "    standard_scale=\"var\",\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    "    colorbar_title=\"Mean expression\",\n",
    "    dendrogram=False,\n",
    "    save=\"TNKILC_annotations_dotplot_mean.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save intermediate result\n",
    "# write to file\n",
    "adata.write(\n",
    "    DIR2SAVE.joinpath(\n",
    "        \"SMC_KUL_Pelka_Che_Wu_CRC_integrated_scvi_hvg_TNKILC_clustering.h5ad\"\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write csv of annotations\n",
    "adata.obs[\"Annotation_scVI_detailed\"].to_csv(\n",
    "    DIR2SAVE.joinpath(\n",
    "        \"SMC_KUL_Pelka_Che_Wu_CRC_integrated_scvi_hvg_annotations_TNKILC.txt\"\n",
    "    ),\n",
    "    sep=\"\\t\",\n",
    "    index=True,\n",
    "    header=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\n",
    "        \"ENTPD1\",\n",
    "        \"CSF1\",\n",
    "        \"KLRC1\",\n",
    "        \"KLRC2\",\n",
    "        \"TIGIT\",\n",
    "        \"CYP26A1\",\n",
    "        \"B4GALNT1\",\n",
    "        \"CD9\",\n",
    "        \"KIR2DL1\",\n",
    "        \"KIR2DL3\",\n",
    "        \"LILRB1\",\n",
    "        \"KLRC3\",\n",
    "    ],\n",
    "    color_map=\"plasma_r\",\n",
    "    use_raw=True,\n",
    ")  # paper dNK1, our NK2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=['APOC1', 'APOE', 'JCHAIN', 'IGHA2', \"CD79A\"], # CD22, CD79A, MS4A1, CD79B APOC1, SPP1, APOE, HSP, IGHG, JCHAIN\n",
    "    use_raw=True,\n",
    "    vmin=0.4,\n",
    "    vmax=\"p99\",\n",
    "    color_map=\"plasma_r\",  #'RdPu',\n",
    "    save=\"general_markers_doublet.pdf\",\n",
    "    show=True,\n",
    ") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_markers_dict = {\n",
    "    \"Myeloid\": ['APOE', 'APOC1'],\n",
    "    \"B\": ['CD79A', 'MS4A1', 'CD22'],\n",
    "    \"T\": ['TRAC', 'CD3E', 'CD3D', 'CD4', 'CD8A'],\n",
    "    'NK': ['NCAM1', 'FCER1G'],\n",
    "    'Cycling': ['MKI67'],\n",
    "    'Treg': ['FOXP3', 'IL2RA'],\n",
    "    'immediate early genes': ['JUN', 'FOS'],\n",
    "    'HSP': [\"HSPA1A\",\n",
    "        \"HSPA1B\",\n",
    "        \"HSP90AA1\",\n",
    "        \"HSP90AB1\",] \n",
    "}\n",
    "\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata,\n",
    "    groupby=\"leiden_scVI\",\n",
    "    key=\"rank_genes_wilcoxon\",\n",
    "    var_names=doublet_markers_dict,\n",
    "    values_to_plot=\"logfoldchanges\",\n",
    "    cmap=\"bwr\",\n",
    "    vmin=-4,\n",
    "    vmax=4,\n",
    "    min_logfoldchange=1,\n",
    "    colorbar_title=\"log fold change\",\n",
    "    save=\"general_dotplot_doublet.pdf\",\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_zscore = adata.raw.to_adata().copy()\n",
    "# scale and store results in layer for later plotting\n",
    "adata_zscore.layers['scaled'] = sc.pp.scale(adata_zscore, copy=True).X\n",
    "adata_zscore.obs=adata.obs\n",
    "\n",
    "\n",
    "sc.pl.dotplot(adata=adata_zscore, \n",
    "              var_names=doublet_markers_dict,\n",
    "              groupby=\"leiden_scVI\",\n",
    "              cmap=\"seismic\", \n",
    "              layer=\"scaled\", \n",
    "              vmin=-2,\n",
    "              vmax=2, \n",
    "              dot_max=0.5, \n",
    "              colorbar_title = \"scaled expression\", \n",
    "              size_title=\"Fraction of cells (%)\",\n",
    "              save='zscore_general_doublet.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=['CD3E', 'CD3G'], # CD22, CD79A, MS4A1, CD79B APOC1, SPP1, APOE, HSP, IGHG, JCHAIN\n",
    "    use_raw=True,\n",
    "    vmin=0,\n",
    "    vmax=\"p99\",\n",
    "    color_map=\"plasma_r\",  #'RdPu',\n",
    "    save=\"general_markers_cd3.pdf\",\n",
    "    show=True,\n",
    ") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cell2loc19",
   "language": "python",
   "name": "cell2loc19"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
